dev/NumberBox/NumberBoxParser.cpp

```diff
@@ -0,0 +1,276 @@
+std::vector<MathToken> NumberBoxParser::GetTokens(const wchar_t* input, winrt::INumberParser numberParser)
+{
+    auto tokens = std::vector<MathToken>();
+
+    bool expectNumber = true;
+    while (input[0] != '\0')
+    {
+        // Skip spaces
+        auto nextChar = input[0];
+        if (nextChar != L' ')
+        {
+            if (expectNumber)
+            {
+                if (nextChar == '(')
+                {
+                    // Open parens are also acceptable, but don't change the next expected token type.
+                    tokens.push_back(MathToken(MathTokenType::Parenthesis, nextChar));
+                }
+                else
+                {
+                    const auto [value, charLength] = GetNextNumber(input, numberParser);
+
+                    if (charLength > 0)
+                    {
+                        tokens.push_back(MathToken(MathTokenType::Numeric, value));
+                        input += charLength - 1; // advance the end of the token
+                        expectNumber = false; // next token should be an operator
+                    }
+                    else
+                    {
+                        return {};
+                    }
+                }
+            }
+            else
+            {
+                if (c_numberBoxOperators.find(nextChar) != std::wstring::npos)
+                {
+                    tokens.push_back(MathToken(MathTokenType::Operator, nextChar));
+                    expectNumber = true; // next token should be a number
+                }
+                else if (nextChar == ')')
+                {
+                    // Closed parens are also acceptable, but don't change the next expected token type.
+                    tokens.push_back(MathToken(MathTokenType::Parenthesis, nextChar));
+                }
+                else
+                {
+                    return {};
+                }
+            }
+        }
+
+        input++;
+    }
+
+    return tokens;
+}
+
+// Attempts to parse a number from the beginning of the given input string. Returns the character size of the matched string.
+std::tuple<double, size_t> NumberBoxParser::GetNextNumber(std::wstring input, winrt::INumberParser numberParser)
+{
+    size_t charLength = 0;
+    double value = std::numeric_limits<double>::quiet_NaN();
+
+    // Attempt to parse anything before an operator or space as a number
+    std::wregex regex(L"^-?([^-+/*\\(\\)\\^\\s]+)");
+    std::wsmatch match;
+    if (std::regex_search(input.cbegin(), input.cend(), match, regex))
+    {
+        // Might be a number
+        const auto matchLength = match[0].length();
+        const auto parsedNum = numberParser.ParseDouble(input.substr(0, matchLength));
+
+        if (parsedNum)
+        {
+            // Parsing was successful
+            value = parsedNum.Value();
+            charLength = matchLength;
+        }
+    }
+
+    return { value, charLength };
```

Comment:
I didn't know this syntax worked!  @jevansaks why have we been using std::make_tuple instead of this?

Reply:
You would use make_tuple when you want to return a tuple type from a function that has an "auto" typed return (such as a lambda). make_tuple has the advantage over directly (or indirectly as in this case) invoking the tuple constructor because you don't have to repeat the type names again. In other words the scenarios are:

```
    std::tuple<double, size_t> y = { value, charLength };
```

```
    auto x = std::make_tuple(value, charLength);
```

And you use make_tuple just because it means you get implicit typing on the template params for tuple so you don't have to type this:
```
   auto z = std::tuple<double, size_t>{ value, charLength };
```

Reply:
hmm, why wouldn't `auto x = {value, charLength};` work? shouldn't the compiler know the types of value and charLength and thus know the type of x as well?

Reply:
Actually it would work but [the compiler says](https://mariusbancila.ro/blog/2017/04/13/cpp17-new-rules-for-auto-deduction-from-braced-init-list/) that the syntax `{ x, y, z }` deduces to `std::initializer_list`, not `std::tuple`.